nautobot/utilities/renderers.py

```diff
@@ -1,15 +1,169 @@
+                valid_actions.append(action)
+            else:
+                invalid_actions.append(action)
+        if invalid_actions:
+            messages.error(request, f"Missing views for action(s) {', '.join(invalid_actions)}")
+        return valid_actions
+
+    def construct_user_permissions(self, request, model):
+        permissions = {}
+        for action in ("add", "change", "delete", "view"):
+            perm_name = get_permission_for_model(model, action)
+            permissions[action] = request.user.has_perm(perm_name)
+        return permissions
+
+    def construct_table(self, view, request, permissions):
+        table = view.table(view.queryset, user=request.user)
+        if "pk" in table.base_columns and (permissions["change"] or permissions["delete"]):
+            table.columns.show("pk")
+
+        # Apply the request context
+        paginate = {
+            "paginator_class": EnhancedPaginator,
+            "per_page": get_paginate_count(request),
+        }
+        return RequestConfig(request, paginate).configure(table)
+
     def get_context(self, data, accepted_media_type, renderer_context):
-        context = super().get_context(data, accepted_media_type, renderer_context)
-        context.update(data)
+        view = renderer_context["view"]
+        request = renderer_context["request"]
+        instance = view.get_object()
+        model = view.queryset.model
+        content_type = ContentType.objects.get_for_model(model)
+        view.queryset = view.filterset(request.GET, view.queryset).qs
+        view.queryset = view.alter_queryset(request)
+        # Compile a dictionary indicating which permissions are available to the current user for this model
+        permissions = self.construct_user_permissions(request, model)
+        # Construct the objects table
+        valid_actions = self.validate_action_buttons(view, request)
+        detail_context = view.get_extra_context(request, "detail", instance)
+        obj = view.alter_obj_for_edit(instance, request, view.args, view.kwargs)
+        form = None
+        table = None
+
+        if data.get("form"):
+            form = data["form"]
+        else:
+            if view.action == "list":
+                table = self.construct_table(view, request, permissions)
+            elif view.action == "destroy":
+                form = ConfirmationForm(initial=request.GET)
+            elif view.action == "create_or_update":
+                initial_data = normalize_querydict(request.GET)
+                form = view.form(instance=obj, initial=initial_data)
+                restrict_form_fields(form, request.user)
+            elif view.action == "bulk_destroy":
+                if request.POST.get("_all"):
+                    if view.bulk_delete_filterset is not None:
+                        pk_list = [
+                            obj.pk for obj in self.bulk_delete_filterset(request.GET, model.objects.only("pk")).qs
```

Comment:
Is there actually a different filterset for `bulk_delete` vs. any other action? Can we get away with using a single `filterset_class`?